File: Sockets/C++/udp_server.cpp
#include <sys/socket.h>
#include <netinet/in.h>
#include <unistd.h>
#include <iostream>
#include <cstring>

int main() {
    int sockfd;
    struct sockaddr_in servaddr, cliaddr;
    char buffer[1024];
    socklen_t len = sizeof(cliaddr);

    sockfd = socket(AF_INET, SOCK_DGRAM, 0);

    servaddr.sin_family = AF_INET;
    servaddr.sin_addr.s_addr = INADDR_ANY;
    servaddr.sin_port = htons(8080);

    bind(sockfd, (struct sockaddr *)&servaddr, sizeof(servaddr));

    std::cout << "UDP Server listening on port 8080..." << std::endl;

    while (1) {
        int n = recvfrom(sockfd, buffer, sizeof(buffer), 0, (struct sockaddr *)&cliaddr, &len);
        buffer[n] = '\0';
        std::cout << "Received: " << buffer << std::endl;

        sendto(sockfd, buffer, n, 0, (struct sockaddr *)&cliaddr, len);
    }

    close(sockfd);
    return 0;
}

File: Sockets/Python/udp_server.py
import socket

# Create UDP socket using socket() - SOCK_DGRAM for UDP
server_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

# Bind to address using bind()
server_socket.bind(('127.0.0.1', 8080))

print("UDP Server listening on port 8080...")

while True:
    # Receive message using recvfrom() - returns data and client address
    data, addr = server_socket.recvfrom(1024)
    print("Received:", data.decode(), "from", addr)

    # Send response using sendto() - sends to specific address
    server_socket.sendto(data, addr)